### PR TITLE
fix(remotecluster): can not upgrade helm release

### LINKED ISSUE #4063: remotecluster can not upgrade helm release
## Bug Report

**What version of Kubernetes are you using?**
v1.23.4
**What version of Chaos Mesh are you using?**
last 

**What did you do? / Minimal Reproducible Example**
update remotecluster can not upgrade helm release

1. Create a remote cluster using the yaml configuration file

```
apiVersion: chaos-mesh.org/v1alpha1
kind: RemoteCluster
metadata:
 name: remote-cls
spec:
 namespace: "chaos-mesh"
 version: "2.5.0"
 configOverride: eyJkYXNoYm9hcmQiOnsiY3JlYXRlIjpmYWxzZX19    //{"dashboard":{"create":false}}
 kubeConfig:
   secretRef:
      name: remote-cls
      namespace: default
      key: remote-cls
```
2.   edit configOverride to another value and apply to k8s
      `configOverride: eyJkYXNoYm9hcmQiOnsiY3JlYXRlIjp0cnVlfX0=`
3. verify remote cluster dashboard is not start

**What did you expect to see?**
expect the remote cluster to be reinstalled
**What did you see instead?**

**Output of chaosctl**

### PR DESCRIPTION
**Pull request by nioshield**
## What problem does this PR solve?

this pr is part of #4063

## What's changed and how it works?
* add `RemoteClusterStatus.ObservedGeneration`  config and save current remotecluster version
* check helm install and upgrade
* fix `RemoteClusterSpec.ConfigOverride` decode value err

## Related changes

- [ ] This change also requires further updates to the [website](https://github.com/chaos-mesh/website) (e.g. docs)
- [ ] This change also requires further updates to the `UI interface`

## Cherry-pick to release branches (optional)

> This PR should be cherry-picked to the following release branches:

- [x] release-2.6
- [ ] release-2.5

### CHANGELOG

> Must include at least one of them.

- [ ] I have updated the `CHANGELOG.md`
- [ ] I have labeled this PR with "no-need-update-changelog"

### Tests

> Must include at least one of them.

- [ ] Unit test
- [ ] E2E test
- [ ] Manual test

### Side effects

- [ ] **Breaking backward compatibility**

## DCO

If you find the DCO check fails, please run commands like below (Depends on the actual situations. For example, if the failed commit isn't the most recent) to fix it:

```shell
git commit --amend --signoff
git push --force
```

### CODE REVIEW

File: pkg/helm/helm_client.go [hunk 1]
```diff
@@ -96,6 +96,20 @@ func (h *HelmClient) UpgradeOrInstall(namespace string, releaseName string, char
 	return result, nil
 }
 
+func (h *HelmClient) UpgradeRelease(namespace string, releaseName string, chart *chart.Chart, values map[string]interface{}) (*release.Release, error) {
```
[the remark this answers is not in the document]
Comment (nioshield) [reply pkg/helm/helm_client.go:99]: updated  test case for this file , please take a look.

File: controllers/multicluster/remotecluster/controller.go [hunk 2]
```diff
@@ -189,26 +201,36 @@ func (r *Reconciler) ensureHelmRelease(ctx context.Context, obj *v1alpha1.Remote
...
-			}
-			release, err = helmClient.UpgradeOrInstall(obj.Spec.Namespace, chaosMeshReleaseName, chart, values)
-			if err != nil {
-				return "", err
-			}
-		} else {
+	values := make(map[string]interface{})
+	if obj.Spec.ConfigOverride != nil {
+		configStr := strings.Trim(string(obj.Spec.ConfigOverride), "\"")
+		configByte, err := base64.StdEncoding.DecodeString(configStr)
```
Comment (g1eny0ung) [inline controllers/multicluster/remotecluster/controller.go:216]: The type of `configOverride` is `json.RawMessage`, not `base64`. I think you're using the one that suits your situation here, but it causes incompatibility. You should revert this change.

**Review by g1eny0ung — CHANGES_REQUESTED** [review]

**Comment by nioshield** [comment]
@g1eny0ung Please take a look

**Comment by nioshield** [comment]
@g1eny0ung Any other suggestions?

**Comment by g1eny0ung** [comment]
> @g1eny0ung Any other suggestions?

@nioshield I still not review the recent changes, so busy last week. 🫨 I will continue to deal with this week.

**Comment by STRRL** [comment]
It seems use base64 encoded string in configOverride not works...

this yaml does not work for me

```yaml
apiVersion: chaos-mesh.org/v1alpha1
kind: RemoteCluster
metadata:
 name: remote-cls
spec:
 namespace: "chaos-mesh"
 version: "2.5.0"
 configOverride: eyJkYXNoYm9hcmQiOnsiY3JlYXRlIjpmYWxzZX19    //{"dashboard":{"create":false}}
 kubeConfig:
   secretRef:
      name: remote-cls
      namespace: default
      key: remote-cls
```

What should I set in field `configOverride` 🤔?

**Comment by nioshield** [comment]
Using base64-encoded strings in configOverride requires special handling of decoded strings, using an unstructured implementation of configOverride, which I'm working on in another branch

**Comment by STRRL** [comment]
> Using base64-encoded strings in configOverride requires special handling of decoded strings, using an unstructured implementation of configOverride, which I'm working on in another branch

Could I use `configOverride` to configure the RemoteCluster with proper value? Or I have to wait the the work.

**Comment by g1eny0ung** [comment]
After referring to https://github.com/kubernetes-sigs/controller-tools/issues/637, I added the appropriate markers and updated `controller-gen` to `v0.6.1`. Following this, I regenerated the CRDs by running `make manifests/crd.yaml`.

After experimenting, I added values to `configOverride` and reapplied. Based on the results, it appears that `json.Unmarshal` is properly handling the changes.

An example:

```yaml
apiVersion: chaos-mesh.org/v1alpha1
kind: RemoteCluster
metadata:
  name: remote-cluster-test
spec:
  namespace: "chaos-mesh"
  version: "2.6.1"
  configOverride:
    dashboard:
      create: false
  kubeConfig:
    secretRef:
      name: secret-kubeconfig
      namespace: default
      key: kubeconfig
```

@nioshield @STRRL Can you help verify this when you have time?

**Comment by nioshield** [comment]
> After referring to [kubernetes-sigs/controller-tools#637](https://github.com/kubernetes-sigs/controller-tools/issues/637), I added the appropriate markers and updated `controller-gen` to `v0.6.1`. Following this, I regenerated the CRDs by running `make manifests/crd.yaml`.
> 
> After experimenting, I added values to `configOverride` and reapplied. Based on the results, it appears that `json.Unmarshal` is properly handling the changes.
> 
> An example:
> 
> ```yaml
> apiVersion: chaos-mesh.org/v1alpha1
> kind: RemoteCluster
> metadata:
>   name: remote-cluster-test
> spec:
>   namespace: "chaos-mesh"
>   version: "2.6.1"
>   configOverride:
>     dashboard:
>       create: false
>   kubeConfig:
>     secretRef:
>       name: secret-kubeconfig
>       namespace: default
>       key: kubeconfig
> ```
> 
> @nioshield @STRRL Can you help verify this when you have time?
I'll check it out later

**Comment by STRRL** [comment]
This PR could not fully resolve the #4063, and we need ongoing changes to fix `configOverride` field.


I would remove the close instruction in the pr's description, and I would review it again in soon

**Comment by STRRL** [comment]
it seems it's required to change the type of `configOverride` to something like

https://github.com/fluxcd/helm-controller/blob/98a5a518a1e717bb379a4f1e7cfc594ea2c78653/api/v2beta1/helmrelease_types.go#L181

**Comment by g1eny0ung** [comment]
> it seems it's required to change the type of `configOverride` to something like
> 
> https://github.com/fluxcd/helm-controller/blob/98a5a518a1e717bb379a4f1e7cfc594ea2c78653/api/v2beta1/helmrelease_types.go#L181-L183

It looks like this also handles `key/value` structures correctly. Have you tried it yourself? 🤔 But I still feel that using these markers is simple and do not need to rely on extra packages:

```go
// +kubebuilder:validation:Schemaless
// +kubebuilder:pruning:PreserveUnknownFields
// +kubebuilder:validation:Type=object
```

**Comment by g1eny0ung** [comment]
> This PR could not fully resolve the #4063, and we need ongoing changes to fix `configOverride` field.
> 
> I would remove the close instruction in the pr's description, and I would review it again in soon

Can you give more details? Is there anything else we need to take care of?

**Comment by nioshield** [comment]
> it seems it's required to change the type of `configOverride` to something like
> 
> https://github.com/fluxcd/helm-controller/blob/98a5a518a1e717bb379a4f1e7cfc594ea2c78653/api/v2beta1/helmrelease_types.go#L181-L183

ConfigOverride is modified in this way and works properly
An example:
```yml
apiVersion: chaos-mesh.org/v1alpha1
kind: RemoteCluster
metadata:
 name: remote-cls
spec:
 namespace: "chaos-mesh"
 version: "2.5.0"
 configOverride:
   dashboard:
      create: false
   controllerManager:
      replicaCount: 2
 kubeConfig:
   secretRef:
      name: remote-cls
      namespace: default
      key: remote-cls
```
@STRRL @g1eny0ung Please look at it again

File: CHANGELOG.md [hunk 3]
```diff
@@ -87,6 +87,7 @@ For more information and how-to, see [RFC: Keep A Changelog](https://github.com/
 - Fix version comparison in install.sh [#3901](https://github.com/chaos-mesh/chaos-mesh/pull/3901)
 - Fix stuck dashboard updates when using ReadWriteOnce PVCs [#3876](https://github.com/chaos-mesh/chaos-mesh/issues/3876)
 - Fix MySQL NO_ZERO_IN_DATE by using `*time.Time` to represent finish time [#4056](https://github.com/chaos-mesh/chaos-mesh/pull/4056)
+- Fix remote cluster cannot upgrade helm release [#4075](https://github.com/chaos-mesh/chaos-mesh/pull/4075)
```
Comment (g1eny0ung) [inline CHANGELOG.md:90]: Good job!

There is one remaining matter that requires attention. Kindly place this line in the unreleased section above. Additionally, there is no need to force-push the changes repeatedly as this results in the loss of context.

**Review by g1eny0ung — CHANGES_REQUESTED** [review]

**Comment by g1eny0ung** [comment]
@nioshield I tried testing the recent changes locally and the results were exactly what I was hoping for! 🥳

But there is a problem with the API server to validate the `configOverride` field if it is passed as a string or anything other than an object. This is due to the absence of an object type restriction.

So I still prefer the solution I found, append below markers and keep the `json.RawMessage` (which also works fine):

```go
// +kubebuilder:validation:Schemaless
// +kubebuilder:validation:Type=object
// +kubebuilder:pruning:PreserveUnknownFields
ConfigOverride json.RawMessage `json:"configOverride,omitempty"`
```

And the solution from @STRRL:

```go
ConfigOverride *apiextensionsv1.JSON `json:"configOverride,omitempty"`
```

I reviewed created CRDs and discovered that using `*apiextensionsv1.JSON` will add `x-kubernetes-preserve-unknown-fields: true` to the OpenAPI definitions. This is equivalent to appending the `// +kubebuilder:pruning:PreserveUnknownFields` marker. If we choose to use `*apiextensionsv1.JSON` as the type of `ConfigOverride`, we still need to add the validation markers described above. I think that would be a bit redundant.

**Comment by nioshield** [comment]
@g1eny0ung Refix and use the markup
```
// +kubebuilder:validation:Schemaless
// +kubebuilder:validation:Type=object
// +kubebuilder:pruning:PreserveUnknownFields
ConfigOverride json.RawMessage `json:"configOverride,omitempty"`
```

**Review by g1eny0ung — APPROVED** [review]
/lgtm

Thanks so much for contributing these past two months! 👍🏼

**Pull request merged**